### PR TITLE
Prevent sstable boundaries from expanding during compaction

### LINKED ISSUE #26: DeleteRange truncation issues
There are a couple issues with range tombstone truncation exposed in these test cases: https://gist.github.com/ajkr/9619663ad7dc5586ac79b29bbc06921f

I believe the current approach (implicit truncation) works nicely for point lookups and iterator next/prev. But the above test cases shows it doesn't handle all cases of iterator seek or writing compaction outputs.

For iterator seek, it should be enough to limit the seek over shadowed data to the sstable boundaries.

For writing compaction outputs it is a bit more difficult. We need to truncate tombstones before writing them but can't simply use the input sstables' boundaries due to cases like https://github.com/facebook/rocksdb/pull/4432#discussion_r221072219. In RocksDB we solved it by truncating to the compaction unit boundary. I believe the analogous thing should work here, even though compaction units are computed differently.

### PR DESCRIPTION
**Pull request by petermattis**
Enforce the invariant that the boundaries of the output of a compaction
can be larger than the boundaries of the input. This fixes a problem
where a compaction can cause a range tombstone to expand to cover a
newer key at a lower level in the LSM.

Fixes #26
Closes #27

### CODE REVIEW

**Comment by petermattis** [comment]
This change is [Reviewable](https://reviewable.io/reviews/petermattis/pebble/67)

**Comment by petermattis** [comment]
This change is so small that I feel I'm missing something. I believe RocksDB achieves the same invariant by truncating range tombstones at read time during compaction. For Pebble I think that truncation is unnecessary as we get implicit truncation to sstable boundaries in `mergingIter`. The part we were getting wrong is allowing the sstable boundaries to expand during compaction. Please give this careful scrutiny. As we're both well aware: range tombstones have super subtle failure scenarios.

File: compaction.go [hunk 1]
```diff
@@ -763,6 +763,25 @@ func (d *DB) compactDiskTables(c *compaction) (ve *versionEdit, pendingOutputs [
 		deletedFiles: map[deletedFileEntry]bool{},
 	}
 
+	// Determine the largest key from the inputs in order to maintain the
+	// invariant that the outputs from a compaction have bounds that are within
+	// the bounds of the input. This is necessary because range tombstones are
+	// not truncated to sstable boundaries. A compaction could this result in the
```
Comment (tbg) [inline compaction.go]: > thus

File: compaction.go [hunk 2]
```diff
@@ -841,6 +860,13 @@ func (d *DB) compactDiskTables(c *compaction) (ve *versionEdit, pendingOutputs [
 		meta.smallest = writerMeta.Smallest(d.cmp)
 		meta.largest = writerMeta.Largest(d.cmp)
 
+		// Bound the largest key of the sstable by the largest key of the
+		// compaction inputs. See the comment on largestInputKey for why this is
+		// necessary.
+		if db.InternalCompare(d.cmp, meta.largest, largestInputKey) > 0 {
```
Comment (tbg) [inline compaction.go]: What allows `meta.largest` to exceed `largestInputKey` in the first place? Is it always a range tombstone?

Comment (tbg) [reply compaction.go]: Maybe the comment on `largestInputKey` could elucidate this more and explain that naively we'd end up with the tombstone which reaches outside of the sstable creating a wider sstable as a result, but we don't want that.

**Comment by ajkr** [comment]
Will try writing some test cases. The cases I'm worried about are when a compaction contains multiple atomic compaction units (in one level). Could be caused by `compaction::grow()` on the input level, or just the chosen input files spanned multiple output level compaction units. In those cases, with this PR I believe we can still write out files whose endpoints and range tombstones extend outside their original compaction unit.

**Comment by petermattis** [comment]
Yeah, there may be a problem here. What if L1 also contained an sstable with the range [d-d)? We’d generate a new sstable in L2 with the bounds [a-d). I’m AFK for a bit, but I can’t see why the approach in this PR would work.

**Comment by petermattis** [comment]
The range tombstones for compaction are loaded by a specially configured levelIter. I bet we could wrap the rangeDelIter to truncate the tombstones to atomic compaction unit boundaries. This would be almost entirely contained in compaction.newInputIter. Let me try that and see how it works.

**Comment by petermattis** [comment]
@ajkr You are right. Here is a slight variation on the test case present in this PR (which was copied from #27) which fails:

```
define target-file-sizes=(100, 1) snapshots=(1)
L0
  a.RANGEDEL.1:e
L0
  a.SET.2:v
L0
  c.SET.3:v
L0
  f.SET.4:v
L2
  d.SET.0:v
----
mem: 1
0: a-e a-a c-c f-f
2: d-d

compact a-b
----
0: f-f
1: a-c c-e
2: d-d

compact d-e
----
0: f-f
1: a-c
3: c-d d-e

get seq=4
c
----
v

compact f-f L0
----
1: a-c f-f
3: c-d d-e

compact a-f L1
----
2: a-e e-f
3: c-d d-e

get seq=4
c
----
v
```

That final compaction should be production L2 sstables with the bounds `a-c` and `f-f`.

**Comment by ajkr** [comment]
Thanks, I was also just trying the same thing you suggested: https://github.com/ajkr/pebble/commit/46464214c8a90a4d3eb6ae666befbfadd7756a37. (Side note: if one accidentally uses `---` instead of `----` in the data-driven tests it's kind of hard to debug.)

**Comment by ajkr** [comment]
> The range tombstones for compaction are loaded by a specially configured levelIter. I bet we could wrap the rangeDelIter to truncate the tombstones to atomic compaction unit boundaries. This would be almost entirely contained in compaction.newInputIter. Let me try that and see how it works.

BTW, I've already written code that pushes atomic compaction unit boundaries down to level iterator. I just hadn't decided yet where to do the actual truncation. Here it is in case any of it is useful: https://github.com/ajkr/pebble/commit/ba9e6a5a6ee564f720ba99655f1bc1947a659407

**Comment by petermattis** [comment]
Heh, I didn't see your messages until just now. I just updated this commit to do the truncation to the atomic compaction unit upper bound by wrapping the rangeDelIter used during compaction. Do we have to truncate both the upper and lower bound, or just the upper bound? I think the lower bound truncation comes for free by the way tombstones are fragmented during compaction.

Let me take a look at your commits

File: compaction.go [hunk 3]
```diff
@@ -267,11 +303,20 @@ func (c *compaction) newInputIter(
 				rangeDelIter = nil
 			}
 		}
+		if rangeDelIter != nil {
+			// Truncate the range tombstones returned by the iterator to the upper
+			// bound of the atomic compaction unit.
+			if upperBound := c.atomicCompactionUnitUpperBound(f); upperBound != nil {
```
Comment (ajkr) [inline compaction.go]: Cool, I didn't think of computing this as the compaction is ongoing.

**Comment by ajkr** [comment]
@petermattis I have a vague memory that old versions of RocksDB don't even truncate their range tombstones if they extend to the left of the file, even though they could because truncating to the start key doesn't lose any information about whether the start key is covered. So maybe. It'd be hard to write a test because the input data would have to be generated by old RocksDB.

File: compaction.go (hunk 3 above)
Comment (petermattis) [reply compaction.go]: Yeah. We could precompute this in `newInputIter` and store the results in a map keyed by `*fileMetadata`, but that seems like overkill when compactions usually involve only a few inputs.

**Comment by petermattis** [comment]
>  I have a vague memory that old versions of RocksDB don't even truncate their range tombstones if they extend to the left of the file, even though they could because truncating to the start key doesn't lose any information about whether the start key is covered. So maybe. It'd be hard to write a test because the input data would have to be generated by old RocksDB.

Ok. I think it is good to be paranoid here and truncate to the `lowerBound` as well. Re testing: we can generate an sstable and then whack it into the LSM in a similar manner to what `runDBDefineCmd` does. 

I'll be AFK the rest of the afternoon and most of tomorrow. Let me know if you want to take over this PR. Otherwise I'll polish it off when I find time.

File: compaction.go [hunk 4]
```diff
@@ -232,6 +232,42 @@ func (c *compaction) elideRangeTombstone(start, end []byte) bool {
...
+//
+// TODO(peter): This needs to be directly tested.
+func (c *compaction) atomicCompactionUnitUpperBound(f *fileMetadata) []byte {
+	for i := range c.inputs {
+		files := c.inputs[i]
+		for j := range files {
+			if f == &files[j] {
+				upperBound := f.largest.UserKey
+				for j = j + 1; j < len(files); j++ {
+					cur := &files[j-1]
```
Comment (ajkr) [inline compaction.go]: Oh yeah, it was also clever to operate just on the compaction input files unlike my implementation, which copied `expandInputs` in that it referenced files in `version`. But that was inconvenient because then we had to worry about holding the DB mutex, I think.

**Comment by petermattis** [comment]
@ajkr I've updated this PR and replaced `truncatingRangeDelIter` with a `rangedel.Truncate` method. This approach is more obviously correct and was somewhat easier to test exhaustively and supports truncation to both the lower and upper boundaries. I've also added a test for the atomic compaction unit boundaries computation. PTAL.

**Comment by petermattis** [comment]
TFTR!
